src/Lib.P2P/Cid.cs

```diff
@@ -0,0 +1,610 @@
+#region LICENSE
+
+/**
+* Copyright (c) 2019 Catalyst Network
+*
+* This file is part of Catalyst.Node <https://github.com/catalyst-network/Catalyst.Node>
+*
+* Catalyst.Node is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.IO;
+using System.Text;
+using Google.Protobuf;
+using MultiFormats;
+using Newtonsoft.Json;
+
+namespace Lib.P2P
+{
+    /// <summary>
+    ///  Identifies some content, e.g. a Content ID.
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///   A Cid is a self-describing content-addressed identifier for distributed systems.
+    ///   </para>
+    ///   <para>
+    ///   Initially, IPFS used a <see cref="MultiHash"/> as the CID and this is still supported as <see cref="Version"/> 0.
+    ///   Version 1 adds a self describing structure to the multi-hash, see the <see href="https://github.com/ipld/cid">spec</see>. 
+    ///   </para>
+    ///   <note>
+    ///   The <see cref="MultiHash.Algorithm">hashing algorithm</see> must be "sha2-256" for a version 0 CID.
+    ///   </note>
+    /// </remarks>
+    /// <seealso href="https://github.com/ipld/cid"/>
+    [JsonConverter(typeof(CidJsonConverter))]
+    public class Cid : IEquatable<Cid>
+    {
+        /// <summary>
+        ///   The default <see cref="ContentType"/>.
+        /// </summary>
+        public const string DefaultContentType = "dag-pb";
+
+        private string encodedValue;
+        private int version;
+        private string encoding = MultiBase.DefaultAlgorithmName;
+        private string contentType = DefaultContentType;
+        private MultiHash hash;
+
+        /// <summary>
+        ///   Throws if a property cannot be set.
+        /// </summary>
+        /// <exception cref="NotSupportedException">
+        ///   When a property cannot be set.
+        /// </exception>
+        /// <remarks>
+        ///   Once <see cref="Encode"/> is invoked, the CID's properties
+        ///   cannot be set.
+        /// </remarks>
+        private void EnsureMutable()
+        {
+            if (encodedValue != null) throw new NotSupportedException("CID cannot be changed.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)